File: Devken.CBC.SchoolManagement.Application/DTOs/Assessments/Assessmentdtos.cs
<<<<<<< HEAD
﻿using System;
using System.ComponentModel.DataAnnotations;
=======
﻿using Devken.CBC.SchoolManagement.Domain.Entities.Assessments;
using Devken.CBC.SchoolManagement.Domain.Enums;
using System.Text.Json.Serialization;
>>>>>>> upstream/main

namespace Devken.CBC.SchoolManagement.Application.DTOs.Assessments
{
    // ─────────────────────────────────────────────────────────────────────────
<<<<<<< HEAD
    // Response DTO
    // ─────────────────────────────────────────────────────────────────────────

    /// <summary>Read-only projection of an Assessment1 entity.</summary>
    public class AssessmentDto
    {
        public Guid Id { get; set; }
        public string Title { get; set; } = null!;
        public string? Description { get; set; }

        public Guid TeacherId { get; set; }
        public string? TeacherName { get; set; }

        public Guid SubjectId { get; set; }
        public string? SubjectName { get; set; }

        public Guid ClassId { get; set; }
        public string? ClassName { get; set; }

        public Guid TermId { get; set; }
        public string? TermName { get; set; }

        public Guid AcademicYearId { get; set; }
        public string? AcademicYearName { get; set; }

        public DateTime AssessmentDate { get; set; }
        public decimal MaximumScore { get; set; }
        public string AssessmentType { get; set; } = null!;  // Formative | Summative | Competency

        public bool IsPublished { get; set; }
        public DateTime? PublishedDate { get; set; }

        public Guid SchoolId { get; set; }
        public DateTime CreatedOn { get; set; }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Create request
    // ─────────────────────────────────────────────────────────────────────────

    public class CreateAssessmentRequest
    {
        [Required, MaxLength(200)]
        public string Title { get; set; } = null!;

        [MaxLength(500)]
        public string? Description { get; set; }

        [Required]
        public Guid TeacherId { get; set; }

        [Required]
        public Guid SubjectId { get; set; }

        [Required]
        public Guid ClassId { get; set; }

        [Required]
        public Guid TermId { get; set; }

        [Required]
        public Guid AcademicYearId { get; set; }

        [Required]
        public DateTime AssessmentDate { get; set; }

        [Required, Range(0.01, 9999.99)]
        public decimal MaximumScore { get; set; }

        /// <summary>Allowed values: Formative, Summative, Competency</summary>
        [Required, MaxLength(20)]
        public string AssessmentType { get; set; } = null!;

        /// <summary>
        /// Required only when the caller is a SuperAdmin.
        /// Regular users inherit their school from the JWT token.
        /// </summary>
        public Guid? SchoolId { get; set; }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Update request
    // ─────────────────────────────────────────────────────────────────────────

    public class UpdateAssessmentRequest
    {
        [Required, MaxLength(200)]
        public string Title { get; set; } = null!;

        [MaxLength(500)]
        public string? Description { get; set; }

        [Required]
        public Guid TeacherId { get; set; }

        [Required]
        public Guid SubjectId { get; set; }

        [Required]
        public Guid ClassId { get; set; }

        [Required]
        public Guid TermId { get; set; }

        [Required]
        public Guid AcademicYearId { get; set; }

        [Required]
        public DateTime AssessmentDate { get; set; }

        [Required, Range(0.01, 9999.99)]
        public decimal MaximumScore { get; set; }

        [Required, MaxLength(20)]
        public string AssessmentType { get; set; } = null!;
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Publish / un-publish request
    // ─────────────────────────────────────────────────────────────────────────

    public class UpdateAssessmentPublishRequest
    {
        /// <summary>True to publish; false to retract.</summary>
        [Required]
        public bool IsPublished { get; set; }
=======
    // ENUMS
    // ─────────────────────────────────────────────────────────────────────────
    public enum AssessmentTypeDto
    {
        Formative = 1,
        Summative = 2,
        Competency = 3
    }

    // ─────────────────────────────────────────────────────────────────────────
    // CREATE / UPDATE REQUEST
    // ─────────────────────────────────────────────────────────────────────────
    public class CreateAssessmentRequest
    {
        // ── Shared ──────────────────────────────────────────────────────
        public AssessmentTypeDto AssessmentType { get; set; }
        public string Title { get; set; } = null!;
        public string? Description { get; set; }
        public Guid TeacherId { get; set; }
        public Guid SubjectId { get; set; }
        public Guid ClassId { get; set; }
        public Guid TermId { get; set; }
        public Guid AcademicYearId { get; set; }
        public DateTime AssessmentDate { get; set; }
        public decimal MaximumScore { get; set; }
        public bool IsPublished { get; set; } = false;
        public Guid? TenantId { get; set; }             // Required for SuperAdmin
        /// <summary>
        /// Frontend alias for TenantId — the Angular form sends "schoolId".
        /// The controller resolves TenantId = TenantId ?? SchoolId before
        /// calling the service, so the service always uses TenantId.
        /// </summary>
        [JsonPropertyName("schoolId")]
        public Guid? SchoolId { get; set; }

        // ── Formative-specific ───────────────────────────────────────────
        public string? FormativeType { get; set; }
        public string? CompetencyArea { get; set; }
        public Guid? LearningOutcomeId { get; set; }
        public string? FormativeStrand { get; set; }
        public string? FormativeSubStrand { get; set; }
        public string? Criteria { get; set; }
        public string? FeedbackTemplate { get; set; }
        public bool RequiresRubric { get; set; } = false;
        public decimal AssessmentWeight { get; set; } = 100.0m;
        public string? FormativeInstructions { get; set; }

        // ── Summative-specific ───────────────────────────────────────────
        public string? ExamType { get; set; }
        public TimeSpan? Duration { get; set; }
        public int NumberOfQuestions { get; set; }
        public decimal PassMark { get; set; } = 50.0m;
        public bool HasPracticalComponent { get; set; } = false;
        public decimal PracticalWeight { get; set; } = 0.0m;
        public decimal TheoryWeight { get; set; } = 100.0m;
        public string? SummativeInstructions { get; set; }

        // ── Competency-specific ──────────────────────────────────────────
        public string? CompetencyName { get; set; }
        public string? CompetencyStrand { get; set; }
        public string? CompetencySubStrand { get; set; }
        public CBCLevel TargetLevel { get; set; }
        public string? PerformanceIndicators { get; set; }
        public AssessmentMethod AssessmentMethod { get; set; }
        public string? RatingScale { get; set; }
        public bool IsObservationBased { get; set; } = true;
        public string? ToolsRequired { get; set; }
        public string? CompetencyInstructions { get; set; }
        public string? SpecificLearningOutcome { get; set; }
    }

    public class UpdateAssessmentRequest : CreateAssessmentRequest
    {
        public Guid Id { get; set; }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // ASSESSMENT RESPONSE
    // ─────────────────────────────────────────────────────────────────────────
    public class AssessmentResponse
    {
        // ── Shared ──────────────────────────────────────────────────────
        public Guid Id { get; set; }
        public AssessmentTypeDto AssessmentType { get; set; }
        public string Title { get; set; } = null!;
        public string? Description { get; set; }
        public Guid TeacherId { get; set; }
        public string TeacherName { get; set; } = null!;
        public Guid SubjectId { get; set; }
        public string SubjectName { get; set; } = null!;
        public Guid ClassId { get; set; }
        public string ClassName { get; set; } = null!;
        public Guid TermId { get; set; }
        public string TermName { get; set; } = null!;
        public Guid AcademicYearId { get; set; }
        public string AcademicYearName { get; set; } = null!;
        public DateTime AssessmentDate { get; set; }
        public decimal MaximumScore { get; set; }
        public bool IsPublished { get; set; }
        public DateTime? PublishedDate { get; set; }
        public DateTime CreatedOn { get; set; }
        public int ScoreCount { get; set; }

        // ── Formative-specific ───────────────────────────────────────────
        public string? FormativeType { get; set; }
        public string? CompetencyArea { get; set; }
        public Guid? LearningOutcomeId { get; set; }
        public string? LearningOutcomeName { get; set; }
        public string? FormativeStrand { get; set; }
        public string? FormativeSubStrand { get; set; }
        public string? Criteria { get; set; }
        public string? FeedbackTemplate { get; set; }
        public bool RequiresRubric { get; set; }
        public decimal AssessmentWeight { get; set; }
        public string? FormativeInstructions { get; set; }

        // ── Summative-specific ───────────────────────────────────────────
        public string? ExamType { get; set; }
        public TimeSpan? Duration { get; set; }
        public int NumberOfQuestions { get; set; }
        public decimal PassMark { get; set; }
        public bool HasPracticalComponent { get; set; }
        public decimal PracticalWeight { get; set; }
        public decimal TheoryWeight { get; set; }
        public string? SummativeInstructions { get; set; }

        // ── Competency-specific ──────────────────────────────────────────
        public string? CompetencyName { get; set; }
        public string? CompetencyStrand { get; set; }
        public string? CompetencySubStrand { get; set; }
        public CBCLevel? TargetLevel { get; set; }
        public string? PerformanceIndicators { get; set; }
        public AssessmentMethod? AssessmentMethod { get; set; }
        public string? RatingScale { get; set; }
        public bool IsObservationBased { get; set; }
        public string? ToolsRequired { get; set; }
        public string? CompetencyInstructions { get; set; }
        public string? SpecificLearningOutcome { get; set; }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // LIST ITEM — lightweight for grid/table views
    // ─────────────────────────────────────────────────────────────────────────
    public class AssessmentListItem
    {
        public Guid Id { get; set; }
        public string Title { get; set; } = null!;
        public AssessmentTypeDto AssessmentType { get; set; }
        public string AssessmentTypeLabel => AssessmentType.ToString();
        public string TeacherName { get; set; } = null!;
        public string SubjectName { get; set; } = null!;
        public string ClassName { get; set; } = null!;
        public string TermName { get; set; } = null!;
        public DateTime AssessmentDate { get; set; }
        public decimal MaximumScore { get; set; }
        public bool IsPublished { get; set; }
        public int ScoreCount { get; set; }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // PUBLISH REQUEST
    // ─────────────────────────────────────────────────────────────────────────
    public class PublishAssessmentRequest
    {
        public AssessmentTypeDto AssessmentType { get; set; }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // SCORE DTOS
    // ─────────────────────────────────────────────────────────────────────────

    /// <summary>
    /// Single request that covers score entry for all three assessment types.
    /// Only populate the fields relevant to the AssessmentType.
    /// </summary>
    public class UpsertScoreRequest
    {
        public Guid? ScoreId { get; set; }                // null = create, set = update
        public AssessmentTypeDto AssessmentType { get; set; }
        public Guid AssessmentId { get; set; }
        public Guid StudentId { get; set; }

        // ── Formative score fields ───────────────────────────────────────
        public decimal? Score { get; set; }
        public decimal? MaximumScore { get; set; }
        public string? Grade { get; set; }
        public string? PerformanceLevel { get; set; }
        public string? Feedback { get; set; }
        public string? Strengths { get; set; }
        public string? AreasForImprovement { get; set; }
        public bool IsSubmitted { get; set; } = false;
        public DateTime? SubmissionDate { get; set; }
        public string? CompetencyArea { get; set; }
        public bool CompetencyAchieved { get; set; } = false;
        public Guid? GradedById { get; set; }

        // ── Summative score fields ───────────────────────────────────────
        public decimal? TheoryScore { get; set; }
        public decimal? PracticalScore { get; set; }
        public decimal? MaximumTheoryScore { get; set; }
        public decimal? MaximumPracticalScore { get; set; }
        public string? Remarks { get; set; }
        public int? PositionInClass { get; set; }
        public int? PositionInStream { get; set; }
        public bool IsPassed { get; set; }
        public string? Comments { get; set; }

        // ── Competency score fields ──────────────────────────────────────
        public string? Rating { get; set; }
        public int? ScoreValue { get; set; }
        public string? Evidence { get; set; }
        public string? AssessmentMethod { get; set; }
        public string? ToolsUsed { get; set; }
        public bool IsFinalized { get; set; } = false;
        public string? Strand { get; set; }
        public string? SubStrand { get; set; }
        public string? SpecificLearningOutcome { get; set; }
        public Guid? AssessorId { get; set; }
    }

    /// <summary>
    /// Unified score response — all fields present, non-relevant ones are null.
    /// The UI reads AssessmentType to know which fields to display.
    /// </summary>
    public class AssessmentScoreResponse
    {
        public Guid Id { get; set; }
        public AssessmentTypeDto AssessmentType { get; set; }
        public Guid AssessmentId { get; set; }
        public string AssessmentTitle { get; set; } = null!;
        public Guid StudentId { get; set; }
        public string StudentName { get; set; } = null!;
        public string StudentAdmissionNo { get; set; } = null!;
        public DateTime AssessmentDate { get; set; }

        // ── Formative ────────────────────────────────────────────────────
        public decimal? Score { get; set; }
        public decimal? MaximumScore { get; set; }
        public decimal? Percentage { get; set; }
        public string? Grade { get; set; }
        public string? PerformanceLevel { get; set; }
        public string? Feedback { get; set; }
        public string? Strengths { get; set; }
        public bool? CompetencyAchieved { get; set; }
        public bool? IsSubmitted { get; set; }
        public string? GradedByName { get; set; }

        // ── Summative ────────────────────────────────────────────────────
        public decimal? TheoryScore { get; set; }
        public decimal? PracticalScore { get; set; }
        public decimal? TotalScore { get; set; }
        public decimal? MaximumTotalScore { get; set; }
        public string? Remarks { get; set; }
        public int? PositionInClass { get; set; }
        public bool? IsPassed { get; set; }
        public string? PerformanceStatus { get; set; }
        public string? Comments { get; set; }

        // ── Competency ───────────────────────────────────────────────────
        public string? Rating { get; set; }
        public string? CompetencyLevel { get; set; }
        public string? Evidence { get; set; }
        public bool? IsFinalized { get; set; }
        public string? AssessorName { get; set; }
        public string? Strand { get; set; }
        public string? SubStrand { get; set; }
>>>>>>> upstream/main
    }
}

File: Devken.CBC.SchoolManagement.API/Devken.CBC.SchoolManagment.UI/School-System-UI/src/app/assessment/types/AssessmentDtos.ts
// ═══════════════════════════════════════════════════════════════════
// AssessmentDtos.ts
// ═══════════════════════════════════════════════════════════════════

// ── Enums ─────────────────────────────────────────────────────────

export enum AssessmentType {
  Formative  = 1,
  Summative  = 2,
  Competency = 3,
}

export enum AssessmentMethod {
  Observation    = 'Observation',
  Portfolio      = 'Portfolio',
  Project        = 'Project',
  Practical      = 'Practical',
  Written        = 'Written',
  Oral           = 'Oral',
  PeerAssessment = 'PeerAssessment',
}

export enum CBCLevel {
  Below      = 'Below',
  Approaching= 'Approaching',
  MeetingExpectations = 'MeetingExpectations',
  Exceeding  = 'Exceeding',
}

// ── Label / colour / icon helpers ─────────────────────────────────

export function getAssessmentTypeLabel(type: AssessmentType): string {
  switch (type) {
    case AssessmentType.Formative:  return 'Formative';
    case AssessmentType.Summative:  return 'Summative';
    case AssessmentType.Competency: return 'Competency';
    default: return 'Unknown';
  }
}

export function getAssessmentTypeColor(type: AssessmentType): string {
  switch (type) {
    case AssessmentType.Formative:  return 'blue';
    case AssessmentType.Summative:  return 'violet';
    case AssessmentType.Competency: return 'pink';
    default: return 'gray';
  }
}

export function getAssessmentTypeIcon(type: AssessmentType): string {
  switch (type) {
    case AssessmentType.Formative:  return 'assignment';
    case AssessmentType.Summative:  return 'school';
    case AssessmentType.Competency: return 'verified_user';
    default: return 'help';
  }
}

// ── Option lists ──────────────────────────────────────────────────

export interface SelectOption { label: string; value: string | number; }

export const AssessmentTypeOptions: SelectOption[] = [
  { label: 'Formative',  value: AssessmentType.Formative  },
  { label: 'Summative',  value: AssessmentType.Summative  },
  { label: 'Competency', value: AssessmentType.Competency },
];

/** Alias kept for components that import ASSESSMENT_TYPES */
export const ASSESSMENT_TYPES = AssessmentTypeOptions;

export const FormativeTypeOptions: SelectOption[] = [
  { label: 'Quiz',              value: 'Quiz'              },
  { label: 'Assignment',        value: 'Assignment'        },
  { label: 'Class Work',        value: 'ClassWork'         },
  { label: 'Homework',          value: 'Homework'          },
  { label: 'Project',           value: 'Project'           },
  { label: 'Practical',         value: 'Practical'         },
  { label: 'Oral Assessment',   value: 'OralAssessment'    },
  { label: 'Portfolio',         value: 'Portfolio'         },
  { label: 'Observation',       value: 'Observation'       },
];

export const ExamTypeOptions: SelectOption[] = [
  { label: 'End of Term Exam',  value: 'EndOfTermExam'     },
  { label: 'Mid Term Exam',     value: 'MidTermExam'       },
  { label: 'Mock Exam',         value: 'MockExam'          },
  { label: 'National Exam',     value: 'NationalExam'      },
  { label: 'Standardised Test', value: 'StandardisedTest'  },
  { label: 'Opening Exam',      value: 'OpeningExam'       },
];

export const RatingScaleOptions: SelectOption[] = [
  { label: 'Exceeds Expectations (EE)', value: 'EE'  },
  { label: 'Meets Expectations (ME)',   value: 'ME'  },
  { label: 'Approaching Expectations (AE)', value: 'AE' },
  { label: 'Below Expectations (BE)',   value: 'BE'  },
];

// ── DTOs ──────────────────────────────────────────────────────────

/** Full assessment response — mirrors C# AssessmentResponse */
export interface AssessmentDto {
  id:               string;
  assessmentType:   AssessmentType;
  title:            string;
  description?:     string;
  teacherId:        string;
  teacherName:      string;
  subjectId:        string;
  subjectName:      string;
  classId:          string;
   schoolId?:        string; 
  className:        string;
  termId:           string;
  termName:         string;
  academicYearId:   string;
  academicYearName: string;
  assessmentDate:   string;
  maximumScore:     number;
  isPublished:      boolean;
  publishedDate?:   string;
  createdOn:        string;
  scoreCount:       number;

  // Formative
  formativeType?:        string;
  competencyArea?:       string;
  strandId?:             string;
  strandName?:           string;
  subStrandId?:          string;
  subStrandName?:        string;
  learningOutcomeId?:    string;
  learningOutcomeName?:  string;
  criteria?:             string;
  feedbackTemplate?:     string;
  requiresRubric?:       boolean;
  assessmentWeight?:     number;
  formativeInstructions?: string;

  // Summative
  examType?:             string;
  duration?:             string;
  numberOfQuestions?:    number;
  passMark?:             number;
  hasPracticalComponent?: boolean;
  practicalWeight?:      number;
  theoryWeight?:         number;
  summativeInstructions?: string;

  // Competency
  competencyName?:          string;
  competencyStrand?:        string;
  competencySubStrand?:     string;
  targetLevel?:             string;
  performanceIndicators?:   string;
  assessmentMethod?:        string;
  ratingScale?:             string;
  isObservationBased?:      boolean;
  toolsRequired?:           string;
  competencyInstructions?:  string;
  specificLearningOutcome?: string;
}

/** Lightweight item for list/grid — mirrors C# AssessmentListItem */
export interface AssessmentListItem {
  id:               string;
  title:            string;
  assessmentType:   AssessmentType;
  assessmentTypeLabel: string;
  teacherName:      string;
  subjectName:      string;
  className:        string;
  termName:         string;
  assessmentDate:   string;
  maximumScore:     number;
  isPublished:      boolean;
  scoreCount:       number;
  strandName?:      string;
  subStrandName?:   string;
}

// ── Request DTOs ──────────────────────────────────────────────────

export interface CreateAssessmentRequest {
  assessmentType:   AssessmentType;
  title:            string;
  description?:     string;
  teacherId:        string;
  subjectId:        string;
  classId:          string;
  termId:           string;
  academicYearId:   string;
  assessmentDate:   string;
  maximumScore:     number;
  isPublished?:     boolean;
  schoolId?:        string;   // frontend alias → backend resolves to tenantId

  // Formative
  formativeType?:        string;
  competencyArea?:       string;
  strandId?:             string;
  subStrandId?:          string;
  learningOutcomeId?:    string;
  criteria?:             string;
  feedbackTemplate?:     string;
  requiresRubric?:       boolean;
  assessmentWeight?:     number;
  formativeInstructions?: string;

  // Summative
  examType?:             string;
  duration?:             string;
  numberOfQuestions?:    number;
  passMark?:             number;
  hasPracticalComponent?: boolean;
  practicalWeight?:      number;
  theoryWeight?:         number;
  summativeInstructions?: string;

  // Competency
  competencyName?:          string;
  competencyStrand?:        string;
  competencySubStrand?:     string;
  targetLevel?:             string;
  performanceIndicators?:   string;
  assessmentMethod?:        string;
  ratingScale?:             string;
  isObservationBased?:      boolean;
  toolsRequired?:           string;
  competencyInstructions?:  string;
  specificLearningOutcome?: string;
}

export interface UpdateAssessmentRequest extends CreateAssessmentRequest {
  id: string;
}

export interface PublishAssessmentRequest {
  assessmentType: AssessmentType;
}
export const ASSESSMENT_TYPE_COLORS: Record<string, string> = {
  Formative:  'blue',
  Summative:  'violet',
  Competency: 'pink',
};

// ── Score DTOs ────────────────────────────────────────────────────

export interface UpsertScoreRequest {
  scoreId?:          string;
  assessmentType:    AssessmentType;
  assessmentId:      string;
  studentId:         string;
  score?:            number;
  maximumScore?:     number;
  grade?:            string;
  performanceLevel?: string;
  feedback?:         string;
  strengths?:        string;
  areasForImprovement?: string;
  isSubmitted?:      boolean;
  submissionDate?:   string;
  competencyArea?:   string;
  competencyAchieved?: boolean;
  gradedById?:       string;
  theoryScore?:      number;
  practicalScore?:   number;
  maximumTheoryScore?: number;
  maximumPracticalScore?: number;
  remarks?:          string;
  positionInClass?:  number;
  positionInStream?: number;
  isPassed?:         boolean;
  comments?:         string;
  rating?:           string;
  scoreValue?:       number;
  evidence?:         string;
  toolsUsed?:        string;
  isFinalized?:      boolean;
  strand?:           string;
  subStrand?:        string;
  specificLearningOutcome?: string;
  assessorId?:       string;
}

export interface AssessmentScoreResponse {
  id:                  string;
  assessmentType:      AssessmentType;
  assessmentId:        string;
  assessmentTitle:     string;
  studentId:           string;
  studentName:         string;
  studentAdmissionNo:  string;
  assessmentDate:      string;
  score?:              number;
  maximumScore?:       number;
  percentage?:         number;
  grade?:              string;
  performanceLevel?:   string;
  feedback?:           string;
  strengths?:          string;
  competencyAchieved?: boolean;
  isSubmitted?:        boolean;
  gradedByName?:       string;
  theoryScore?:        number;
  practicalScore?:     number;
  totalScore?:         number;
  maximumTotalScore?:  number;
  remarks?:            string;
  positionInClass?:    number;
  isPassed?:           boolean;
  performanceStatus?:  string;
  comments?:           string;
  rating?:             string;
  competencyLevel?:    string;
  evidence?:           string;
  isFinalized?:        boolean;
  assessorName?:       string;
  strand?:             string;
  subStrand?:          string;
}

// After AssessmentMethod enum
export const AssessmentMethodOptions: SelectOption[] = [
  { label: 'Observation',      value: AssessmentMethod.Observation    },
  { label: 'Portfolio',        value: AssessmentMethod.Portfolio       },
  { label: 'Project',          value: AssessmentMethod.Project         },
  { label: 'Practical',        value: AssessmentMethod.Practical       },
  { label: 'Written',          value: AssessmentMethod.Written         },
  { label: 'Oral',             value: AssessmentMethod.Oral            },
  { label: 'Peer Assessment',  value: AssessmentMethod.PeerAssessment  },
];